qcore/gsf.py

```diff
@@ -109,55 +111,58 @@ def coordinate_meshgrid(
     ).T
     nztm_meshgrid = (transformation_matrix @ subdivision_coordinates).T
     nztm_meshgrid += origin
-    return coordinates.nztm_to_wgs_depth(nztm_meshgrid)
+    return coordinates.nztm_to_wgs_depth(nztm_meshgrid).reshape((ny, nx, 3))
 
 
 def write_fault_to_gsf_file(
     gsf_filepath: Path,
-    meshgrids: list[np.ndarray],
-    lengths: np.ndarray,
-    widths: np.ndarray,
-    strikes: np.ndarray,
-    dips: np.ndarray,
-    rakes: np.ndarray,
+    gsf_df: pd.DataFrame,
     resolution: int,
 ):
     """Writes geometry data to a GSF file.
 
+    This code assumes that the dip is constant across all faults.
+
     Parameters
     ----------
     gsf_filepath : Path
         The file path pointing to the GSF file to write to.
-    meshgrids : list[np.ndarray]
-        List of meshgrid arrays representing fault segments (length: n).
-    lengths : np.ndarray
-        An (n x 1) array of segment lengths.
-    widths : np.ndarray
-        An (n x 1) array of segment widths.
-    strikes : np.ndarray
-        An (n x 1) array of segment strikes.
-    dips : np.ndarray
-        An (n x 1) array of segment dips.
-    rakes : np.ndarray
-        An (n x 1) array of segment rakes.
+    gsf_df : pd.DataFrame
+        The GSF dataframe to write. This dataframe must have the columns length,
+        width, strike, dip, rake, and meshgrid. Each row corresponds to one
+        fault plane, with the meshgrid column being the discretisation of the
+        fault planes.
     resolution : int
         Resolution of the meshgrid.
     """
     with open(gsf_filepath, "w", encoding="utf-8") as gsf_file_handle:
         gsf_file_handle.write(
             "# LON  LAT  DEP(km)  SUB_DX  SUB_DY  LOC_STK  LOC_DIP  LOC_RAKE  SLIP(cm)  INIT_TIME  SEG_NO\n"
         )
-        number_of_points = sum(meshgrid.shape[0] for meshgrid in meshgrids)
+        number_of_points = gsf_df.apply(
+            lambda row: np.prod(row["meshgrid"].shape[:2]), axis=1
+        ).sum()
+
+        # Get the number of dip gridpoints by looking at the first dimension of
+        # the meshgrid of the first fault plane. See coordinate_meshgrid for an
+        # explanation of meshgrid dimensions.
+        n_dip = gsf_df.iloc[0]['meshgrid'].shape[0]
+
         gsf_file_handle.write(f"{number_of_points}\n")
-        for i, (length, width, strike, dip, rake, meshgrid) in enumerate(
-            zip(lengths, widths, strikes, dips, rakes, meshgrids)
-        ):
-            strike_step = length / gridpoint_count_in_length(length * 1000, resolution)
-            dip_step = width / gridpoint_count_in_length(width * 1000, resolution)
-            for point in meshgrid:
-                gsf_file_handle.write(
-                    f"{point[1]:11.5f} {point[0]:11.5f} {point[2] / 1000:11.5e} {strike_step:11.5e} {dip_step:11.5e} {strike:6.1f} {dip:6.1f} {rake:6.1f} {-1.0:8.2f} {-1.0:8.2f} {i:3d}\n"
-                )
+        for j in range(n_dip):
```

Comment:
Not sure how I feel about the variable j. Maybe dip_index or depth_index? As it's the index as you go down the dip / go deeper. Maybe even i could change to plane_index